### PR TITLE
Make the private functions of the new return system public // Documentation

### PR DESCRIPTION
**Pull request by rmoyard**
**Description of the Change:**

All the execute from the new return system replace the legacy system, therefore the new functions become public and the old system one become private. The new return system documentation therefore become user facing.

Replace `mode` by `execution_on_grad` in the tests.

### CODE REVIEW

File: pennylane/interfaces/torch.py [hunk 1]
```diff
@@ -497,7 +495,7 @@ def _execute_new(tapes, device, execute_fn, gradient_fn, gradient_kwargs, _n=1,
         "max_diff": max_diff,
     }
 
-    return ExecuteTapesNew.apply(kwargs, *parameters)
+    return ExecuteTapes.apply(kwargs, *parameters)
```
Comment (mudit2812) [inline pennylane/interfaces/torch.py:500]: Change `_compute_vjps` and `_compute_vjps_new` in this file as well.

File: pennylane/transforms/qcut/montecarlo.py [hunk 2]
```diff
@@ -517,6 +517,11 @@ def _wrapper(*args, **kwargs):
 
         execute_kwargs["cache"] = False
 
+        if "mode" in execute_kwargs:
+            mode = execute_kwargs.pop("mode")
+            if not qml.active_return():
+                execute_kwargs["grad_on_execution"] = "forward" if mode != "best" else "best"
```
Comment (mudit2812) [inline pennylane/transforms/qcut/montecarlo.py:523]: Do we only use `grad_on_execution` when the new return types are disabled?

File: tests/interfaces/test_autograd_new.py [hunk 3]
```diff
@@ -24,7 +24,6 @@
 import pennylane as qml
 from pennylane.devices import DefaultQubit
 from pennylane.gradients import finite_diff, param_shift
-from pennylane.interfaces import execute
```
Comment (mudit2812) [inline tests/interfaces/test_autograd_new.py:27]: Should this test file be renamed to `test_autograd.py`?

File: tests/interfaces/test_autograd_qnode_new.py [hunk 4]
```diff
@@ -22,30 +22,30 @@
 from pennylane.tape import QuantumScript
```
Comment (mudit2812) [inline tests/interfaces/test_autograd_qnode_new.py:23]: Same question for this file.

File: tests/interfaces/test_torch_qnode.py [hunk 5]
```diff
@@ -1644,7 +1730,7 @@ def circuit():
         assert result[2].dtype == torch.int64
 
 
-qubit_device_and_diff_method_and_mode = [
+qubit_device_and_diff_method_and_grad_on_execution = [
     ["default.qubit", "backprop", "forward"],
     ["default.qubit", "finite-diff", "backward"],
     ["default.qubit", "parameter-shift", "backward"],
```
Comment (mudit2812) [inline tests/interfaces/test_torch_qnode.py:1736]: The `grad_on_execution` would make more sense here to be booleans rather than `"forward"` or `"backward"`

File: tests/legacy/test_jax_old.py [hunk 6]
```diff
@@ -369,7 +369,7 @@ def cost(a, cache):
                 dev,
                 gradient_fn="device",
                 cache=cache,
-                mode="backward",
+                grad_on_execution="backward",
```
Comment (mudit2812) [inline tests/legacy/test_jax_old.py:372]:
```suggestion
                grad_on_execution=False,
```

File: tests/legacy/test_jax_old.py [hunk 7]
```diff
@@ -95,9 +95,9 @@ def cost(a, device):
         for args in spy.call_args_list:
             assert args[1]["shifts"] == [(np.pi / 4,)] * 2
 
-    def test_incorrect_mode(self, interface):
+    def test_incorrect_grad_on_execution(self, interface):
         """Test that an error is raised if an gradient transform
-        is used with mode=forward"""
+        is used with grad_on_execution=forward"""
```
Comment (mudit2812) [inline tests/legacy/test_jax_old.py:100]:
```suggestion
        is used with grad_on_execution=True"""
```

**Review by mudit2812 — CHANGES_REQUESTED** [review]
Thanks for the changes Romain! This PR is looking pretty good. I found a few artefacts referencing the `_new` version of functions, which I've included in my review.

File: tests/legacy/test_jax_old.py (hunk 7 above)
Comment (rmoyard) [reply tests/legacy/test_jax_old.py:100]: I kept mode for the old return type.

File: tests/legacy/test_jax_old.py [hunk 8]
```diff
@@ -352,7 +352,7 @@ def cost(a, cache):
 
     def test_caching_adjoint_backward(self, interface):
         """Test that caching produces the optimum number of adjoint evaluations
-        when mode=backward"""
+        when grad_on_execution=backward"""
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply tests/legacy/test_jax_old.py:355]: I kept mode for the old return type.

File: tests/legacy/test_jax_old.py (hunk 6 above)
Comment (rmoyard) [reply tests/legacy/test_jax_old.py:372]: I kept mode for the old return type.

File: pennylane/transforms/qcut/montecarlo.py (hunk 2 above)
Comment (rmoyard) [reply pennylane/transforms/qcut/montecarlo.py:523]: The grad on exeution is only given with the new return system so here I convert  mode to the grad on execution.

File: tests/interfaces/test_torch_qnode.py (hunk 5 above)
Comment (rmoyard) [reply tests/interfaces/test_torch_qnode.py:1736]: Good catch I missed it!.

File: tests/test_qubit_device.py [hunk 9]
```diff
@@ -326,31 +326,6 @@ def test_parameters_accessed_outside_execution_context(self, mock_qubit_device):
 class TestExtractStatistics:
     """Test the statistics method"""
 
-    def test_observables_deprecated(self, mock_qubit_device_extract_stats, monkeypatch):
```
Comment (lillian542) [inline tests/test_qubit_device.py:329]: Is this related to the return types?

File: tests/interfaces/test_jax_qnode.py [hunk 10]
```diff
@@ -827,21 +859,21 @@ def cost_fn(a, b):
...
+@pytest.mark.parametrize("dev_name,diff_method,grad_on_execution", qubit_device_and_diff_method)
 class TestQubitIntegration:
     """Tests that ensure various qubit circuits integrate correctly"""
 
-    def test_sampling(self, dev_name, diff_method, mode):
+    def test_sampling(self, dev_name, diff_method, grad_on_execution):
         """Test sampling works as expected"""
-        if mode == "forward":
-            pytest.skip("Sampling not possible with forward mode differentiation.")
+        if grad_on_execution:
```
Comment (lillian542) [inline tests/interfaces/test_jax_qnode.py:868]: Could this be "best"? And if so, should this test be skipped?

File: tests/interfaces/test_jax_qnode.py [hunk 11]
```diff
@@ -856,17 +888,17 @@ def circuit():
         assert isinstance(res[1], jax.Array)
         assert res[1].shape == (10,)
 
-    def test_counts(self, dev_name, diff_method, mode):
+    def test_counts(self, dev_name, diff_method, grad_on_execution):
         """Test counts works as expected"""
-        if mode == "forward":
-            pytest.skip("Sampling not possible with forward mode differentiation.")
+        if grad_on_execution:
```
Comment (lillian542) [inline tests/interfaces/test_jax_qnode.py:893]: Same as above

**Review by lillian542 — COMMENTED** [review]
Looks good! Left a few small comments and questions.

File: pennylane/qnode.py [hunk 12]
```diff
@@ -127,6 +127,10 @@ class QNode:
             executed on a device. Expansion occurs when an operation or measurement is not
             supported, and results in a gate decomposition. If any operations in the decomposition
             remain unsupported by the device, another expansion occurs.
+        grad_on_execution (str): (bool, str): Whether the gradients should be computed on the execution or not.
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/qnode.py:130]: Good catch thanks!

File: tests/interfaces/test_autograd_qnode_new.py [hunk 13]
```diff
@@ -1658,7 +1752,7 @@ class TestSample:
     """Tests for the sample integration"""
 
     def test_backprop_error(self):
-        """Test that sampling in backpropagation mode raises an error"""
+        """Test that sampling in backpropagation grad_on_execution raises an error"""
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply tests/interfaces/test_autograd_qnode_new.py:1755]: thanks!

File: tests/test_qubit_device.py (hunk 9 above)
Comment (rmoyard) [reply tests/test_qubit_device.py:329]: Yes we never added observables to statistics, so no deprecation here.

File: tests/interfaces/test_jax_qnode.py (hunk 10 above)
Comment (rmoyard) [reply tests/interfaces/test_jax_qnode.py:868]: We don't use best in the test.

Comment (rmoyard) [reply tests/interfaces/test_jax_qnode.py]: I updated it in the new tests wiht `is True`

File: tests/interfaces/test_jax_qnode.py (hunk 11 above)
Comment (rmoyard) [reply tests/interfaces/test_jax_qnode.py]: I updated it in the new tests wiht `is True`

**Review by mudit2812 — APPROVED** [review]
Looks good to me :)

**Review by lillian542 — APPROVED** [review]
Looks good to me! 🚀
